internal/cri/server/images/image_pull.go

```diff
@@ -798,3 +838,182 @@ func (c *CRIImageService) snapshotterFromPodSandboxConfig(ctx context.Context, i
 
 	return snapshotter, nil
 }
+
+type criCredentials struct {
+	ref         string
+	credentials func(string) (string, string, error)
+}
+
+func newcriCredentials(ref string, credentials func(string) (string, string, error)) registry.CredentialHelper {
+	return &criCredentials{
+		ref:         ref,
+		credentials: credentials,
+	}
+}
+
+// GetCredentials gets credential from criCredentials makes criCredentials a registry.CredentialHelper
+func (cc *criCredentials) GetCredentials(ctx context.Context, ref string, host string) (registry.Credentials, error) {
+	if ref != cc.ref || cc.credentials == nil {
+		return registry.Credentials{}, nil
+	}
+
+	username, secret, err := cc.credentials(host)
+	if err != nil {
+		return registry.Credentials{}, nil
```

Comment:
Shouldn't we return the actual err here?

Reply:
Yes, I have added errors instead of returning nil.